services/skus/credentials.go

```diff
@@ -242,3 +462,211 @@ var generateCredentialRedemptions = func(ctx context.Context, cb []CredentialBin
 	}
 	return requestCredentials, nil
 }
+
+// CredentialBinding includes info needed to redeem a single credential
+type CredentialBinding struct {
+	PublicKey     string `json:"publicKey" valid:"base64"`
+	TokenPreimage string `json:"t" valid:"base64"`
+	Signature     string `json:"signature" valid:"base64"`
+}
+
+// deduplicateCredentialBindings - given a list of tokens return a deduplicated list
+func deduplicateCredentialBindings(tokens ...CredentialBinding) []CredentialBinding {
+	var (
+		seen   = map[string]bool{}
+		result []CredentialBinding
+	)
+	for _, t := range tokens {
+		if !seen[t.TokenPreimage] {
+			seen[t.TokenPreimage] = true
+			result = append(result, t)
+		}
+	}
+	return result
+}
+
+func decodeIssuerID(issuerID string) (string, string, error) {
+	var (
+		merchantID string
+		sku        string
+	)
+
+	u, err := url.Parse(issuerID)
+	if err != nil {
+		return "", "", fmt.Errorf("parse issuer name: %w", err)
+	}
+
+	sku = u.Query().Get("sku")
+	u.RawQuery = ""
+	merchantID = u.String()
+
+	return merchantID, sku, nil
+}
+
+func encodeIssuerID(merchantID, sku string) (string, error) {
+	v := url.Values{}
+	v.Add("sku", sku)
+
+	u, err := url.Parse(merchantID + "?" + v.Encode())
+	if err != nil {
+		return "", fmt.Errorf("parse merchant id: %w", err)
+	}
+
+	return u.String(), nil
+}
+
+// SignedOrderCredentialsHandler - this is the handler for getting the signed order credentials
+type SignedOrderCredentialsHandler struct {
+	decoder   Decoder
+	datastore Datastore
+}
+
+// Handle processes Kafka message of type SigningOrderResult.
+func (s *SignedOrderCredentialsHandler) Handle(ctx context.Context, message kafka.Message) (err error) {
+	signedOrderResult, err := s.decoder.Decode(message)
+	if err != nil {
+		return fmt.Errorf("error decoding message key %s partition %d offset %d: %w",
+			string(message.Key), message.Partition, message.Offset, err)
+	}
+
+	requestID, err := uuid.FromString(signedOrderResult.RequestID)
+	if err != nil {
+		return fmt.Errorf("error getting uuid from signed order request %w", err)
+	}
+
+	// Use the outbox message request id to get a session lock on that id to prevent any updates for that outbox message.
```

Comment:
We are trying to prevent in the unlikely event there is a duplicate message (based on request id) being processed at the same time.  Pretty rare edge case I think.

Reply:
Its not so much a race condition but deadlock, this was surfaced during a soak test, to be on the safe side an advisory lock was added which also prevents unnecessary rollbacks for duplicates. There is a TODO (which I removed from the comments) to carry out deeper investigation into this but didn't want to block the PR. Now the constraint have also changed with the request_id the behaviour may be different again.

Reply:
> We are trying to prevent in the unlikely event there is a duplicate message (based on request id) being processed at the same time. Pretty rare edge case I think.

yes, the lock also prevents Kafka and request duplicates being processed at the same time, these would just be rolled back. It is most likely this that caused errors in pg

Reply:
updated handler to reflect addition in request id